arviz/stats/diagnostics.py

```diff
@@ -67,65 +65,44 @@ def _ess_ufunc(ary):
     This can be used on an xarray Dataset, using
     `xr.apply_ufunc(_ess_ufunc, ..., input_core_dims=(('chain', 'draw'),))
     """
-    target = np.empty(ary.shape[:-2])
-    for idx in np.ndindex(target.shape):
-        target[idx] = _get_ess(ary[idx])
-    return target
+    return _get_ess(ary)
 
 
 def _get_ess(sample_array):
-    """Compute the effective sample size for a 2D array."""
-    shape = sample_array.shape
-    if len(shape) != 2:
-        raise TypeError("Effective sample size calculation requires 2 dimensional arrays.")
-    n_chain, n_draws = shape
-    if n_chain <= 1:
-        raise TypeError("Effective sample size calculation requires multiple chains.")
+    """Compute the effective sample size for a ND array with the last two dimensions are chain
+    dimension and draw dimension.
+    """
+    if sample_array.ndim == 1:
+        sample_array = sample_array[None, ...]
+    n_chain, n_draws = sample_array.shape[-2], sample_array.shape[-1]
 
-    acov = np.asarray([_autocov(sample_array[chain]) for chain in range(n_chain)])
+    acov = _autocov(sample_array)
 
-    chain_mean = sample_array.mean(axis=1)
-    chain_var = acov[:, 0] * n_draws / (n_draws - 1.0)
-    acov_t = acov[:, 1] * n_draws / (n_draws - 1.0)
-    mean_var = np.mean(chain_var)
+    chain_mean = sample_array.mean(axis=-1, keepdims=True)
+    chain_var = acov[..., :1] * n_draws / (n_draws - 1.0)
+    mean_var = np.mean(chain_var, axis=-2)
     var_plus = mean_var * (n_draws - 1.0) / n_draws
-    var_plus += np.var(chain_mean, ddof=1)
-
-    rho_hat_t = np.zeros(n_draws)
-    rho_hat_even = 1.0
-    rho_hat_t[0] = rho_hat_even
-    rho_hat_odd = 1.0 - (mean_var - np.mean(acov_t)) / var_plus
-    rho_hat_t[1] = rho_hat_odd
+    if n_chain > 1:
+        var_plus += np.var(chain_mean, axis=-2, ddof=1)
 
     # Geyer's initial positive sequence
-    max_t = 1
-    t = 1
-    while t < (n_draws - 2) and (rho_hat_even + rho_hat_odd) >= 0.0:
-        rho_hat_even = 1.0 - (mean_var - np.mean(acov[:, t + 1])) / var_plus
-        rho_hat_odd = 1.0 - (mean_var - np.mean(acov[:, t + 2])) / var_plus
```

Comment:
while we don't multiply with `n_draws / (n_draws - 1.0)` for the remaining value

Reply:
This behaviour is from the ess paper.

Reply:
Yup, this is correct to me. I just pointed out the inconsistent of current code, to explain for why there are a tiny difference between the current result and this PR result.